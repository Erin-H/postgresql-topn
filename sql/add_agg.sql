CREATE EXTENSION topn;

--
--Testing topn_add_agg function of the extension
--

--prepare tables for aggregates
create table numbers (
	int_column int
);

INSERT INTO numbers SELECT NULL FROM generate_series(1,5);
INSERT INTO numbers SELECT 0 FROM generate_series(1,2);
INSERT INTO numbers SELECT 1 FROM generate_series(1,1);
INSERT INTO numbers SELECT 2 FROM generate_series(1,6);
INSERT INTO numbers SELECT 3 FROM generate_series(1,4);
INSERT INTO numbers SELECT 4 FROM generate_series(1,3);
INSERT INTO numbers SELECT 5 FROM generate_series(1,7);

create table strings (
	text_column text
);

INSERT INTO strings SELECT NULL FROM generate_series(1,30);
INSERT INTO strings SELECT '0' FROM generate_series(1,2);
INSERT INTO strings SELECT '1' FROM generate_series(1,1);
INSERT INTO strings SELECT '2' FROM generate_series(1,15000);
INSERT INTO strings SELECT '3' FROM generate_series(1,20);
INSERT INTO strings SELECT '4' FROM generate_series(1,6);
INSERT INTO strings SELECT '5' FROM generate_series(1,70000);

create table cidr_table (
	cidr_column cidr
);

INSERT INTO cidr_table SELECT NULL FROM generate_series(1,30);
INSERT INTO cidr_table SELECT '192.168.100.128/25'::cidr FROM generate_series(1,2);
INSERT INTO cidr_table SELECT NULL::cidr FROM generate_series(1,1);
INSERT INTO cidr_table SELECT '192.168/24'::cidr FROM generate_series(1,15000);
INSERT INTO cidr_table SELECT '192.168/25'::cidr FROM generate_series(1,20);
INSERT INTO cidr_table SELECT '192.168.1'::cidr FROM generate_series(1,6);
INSERT INTO cidr_table SELECT '192.168.5'::cidr FROM generate_series(1,70000);

create table inet_table (
	inet_column inet
);

INSERT INTO inet_table SELECT NULL FROM generate_series(1,30);
INSERT INTO inet_table SELECT '192.168.100.128/25'::inet FROM generate_series(1,2);
INSERT INTO inet_table SELECT NULL::inet FROM generate_series(1,1);
INSERT INTO inet_table SELECT '192.168.2.1/24'::inet FROM generate_series(1,15000);
INSERT INTO inet_table SELECT '192.168.100.128/23'::inet FROM generate_series(1,20);
INSERT INTO inet_table SELECT '192.168.2.1'::inet FROM generate_series(1,6);
INSERT INTO inet_table SELECT '10.1.2.3/32'::inet FROM generate_series(1,70000);

--Test invalid parameters
SET topn.number_of_counters to 0;
SET topn.number_of_counters to -1;
SET topn.number_of_counters to 1000000000000;

SET topn.number_of_counters to 4;
--check aggregates for fixed size types like integer
SELECT topn(topn_add_agg(int_column::text), 0) FROM numbers WHERE int_column < 0;
SELECT topn(topn_add_agg(int_column::text), 1) FROM numbers WHERE int_column < 1;
SELECT topn(topn_add_agg(int_column::text), 2) FROM numbers WHERE int_column < 2;
SELECT topn(topn_add_agg(int_column::text), 3) FROM numbers WHERE int_column < 3;
SELECT topn(topn_add_agg(int_column::text), 4) FROM numbers WHERE int_column < 4;
SELECT topn(topn_add_agg(int_column::text), 4) FROM numbers WHERE int_column < 5;
SELECT topn(topn_add_agg(int_column::text), 4) FROM numbers WHERE int_column < 6;
SELECT topn(topn_add_agg(int_column::text), 4) FROM numbers;

--check aggregates for variable length types like text
SELECT topn(topn_add_agg(text_column), 0) FROM strings WHERE text_column < '0';
SELECT topn(topn_add_agg(text_column), 1) FROM strings WHERE text_column < '1';
SELECT topn(topn_add_agg(text_column), 2) FROM strings WHERE text_column < '2';
SELECT topn(topn_add_agg(text_column), 3) FROM strings WHERE text_column < '3';
SELECT topn(topn_add_agg(text_column), 4) FROM strings WHERE text_column < '4';
SELECT topn(topn_add_agg(text_column), 4) FROM strings WHERE text_column < '5';
SELECT topn(topn_add_agg(text_column), 4) FROM strings WHERE text_column < '6';
SELECT topn(topn_add_agg(text_column), 4) FROM strings;

--check aggregates for cidr type
SELECT topn(topn_add_agg(cidr_column::text), 4) FROM cidr_table;

--check aggregates for inet type
SELECT topn(topn_add_agg(inet_column::text), 4) FROM inet_table;
