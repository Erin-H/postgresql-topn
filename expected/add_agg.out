CREATE EXTENSION topn;
--
--Testing topn_add_agg function of the extension
--
--prepare tables for aggregates
create table numbers (
	int_column int
);
INSERT INTO numbers SELECT NULL FROM generate_series(1,5);
INSERT INTO numbers SELECT 0 FROM generate_series(1,2);
INSERT INTO numbers SELECT 1 FROM generate_series(1,1);
INSERT INTO numbers SELECT 2 FROM generate_series(1,6);
INSERT INTO numbers SELECT 3 FROM generate_series(1,4);
INSERT INTO numbers SELECT 4 FROM generate_series(1,3);
INSERT INTO numbers SELECT 5 FROM generate_series(1,7);
create table strings (
	text_column text
);
INSERT INTO strings SELECT NULL FROM generate_series(1,30);
INSERT INTO strings SELECT '0' FROM generate_series(1,2);
INSERT INTO strings SELECT '1' FROM generate_series(1,1);
INSERT INTO strings SELECT '2' FROM generate_series(1,15000);
INSERT INTO strings SELECT '3' FROM generate_series(1,20);
INSERT INTO strings SELECT '4' FROM generate_series(1,6);
INSERT INTO strings SELECT '5' FROM generate_series(1,70000);
create table cidr_table (
	cidr_column cidr
);
INSERT INTO cidr_table SELECT NULL FROM generate_series(1,30);
INSERT INTO cidr_table SELECT '192.168.100.128/25'::cidr FROM generate_series(1,2);
INSERT INTO cidr_table SELECT NULL::cidr FROM generate_series(1,1);
INSERT INTO cidr_table SELECT '192.168/24'::cidr FROM generate_series(1,15000);
INSERT INTO cidr_table SELECT '192.168/25'::cidr FROM generate_series(1,20);
INSERT INTO cidr_table SELECT '192.168.1'::cidr FROM generate_series(1,6);
INSERT INTO cidr_table SELECT '192.168.5'::cidr FROM generate_series(1,70000);
create table inet_table (
	inet_column inet
);
INSERT INTO inet_table SELECT NULL FROM generate_series(1,30);
INSERT INTO inet_table SELECT '192.168.100.128/25'::inet FROM generate_series(1,2);
INSERT INTO inet_table SELECT NULL::inet FROM generate_series(1,1);
INSERT INTO inet_table SELECT '192.168.2.1/24'::inet FROM generate_series(1,15000);
INSERT INTO inet_table SELECT '192.168.100.128/23'::inet FROM generate_series(1,20);
INSERT INTO inet_table SELECT '192.168.2.1'::inet FROM generate_series(1,6);
INSERT INTO inet_table SELECT '10.1.2.3/32'::inet FROM generate_series(1,70000);
--Test invalid parameters
SET topn.number_of_counters to 0;
ERROR:  0 is outside the valid range for parameter "topn.number_of_counters" (1 .. 14913080)
SET topn.number_of_counters to -1;
ERROR:  -1 is outside the valid range for parameter "topn.number_of_counters" (1 .. 14913080)
SET topn.number_of_counters to 1000000000000;
ERROR:  invalid value for parameter "topn.number_of_counters": "1000000000000"
HINT:  Value exceeds integer range.
SET topn.number_of_counters to 4;
--check aggregates for fixed size types like integer
SELECT topn(topn_add_agg(int_column::text), 0) FROM numbers WHERE int_column < 0;
 topn 
------
(0 rows)

SELECT topn(topn_add_agg(int_column::text), 1) FROM numbers WHERE int_column < 1;
 topn  
-------
 (0,2)
(1 row)

SELECT topn(topn_add_agg(int_column::text), 2) FROM numbers WHERE int_column < 2;
 topn  
-------
 (0,2)
 (1,1)
(2 rows)

SELECT topn(topn_add_agg(int_column::text), 3) FROM numbers WHERE int_column < 3;
 topn  
-------
 (2,6)
 (0,2)
 (1,1)
(3 rows)

SELECT topn(topn_add_agg(int_column::text), 4) FROM numbers WHERE int_column < 4;
 topn  
-------
 (2,6)
 (3,4)
 (0,2)
 (1,1)
(4 rows)

SELECT topn(topn_add_agg(int_column::text), 4) FROM numbers WHERE int_column < 5;
 topn  
-------
 (2,6)
 (3,4)
 (4,3)
 (0,2)
(4 rows)

SELECT topn(topn_add_agg(int_column::text), 4) FROM numbers WHERE int_column < 6;
 topn  
-------
 (5,7)
 (2,6)
 (3,4)
 (4,3)
(4 rows)

SELECT topn(topn_add_agg(int_column::text), 4) FROM numbers;
 topn  
-------
 (5,7)
 (2,6)
 (3,4)
 (4,3)
(4 rows)

--check aggregates for fixed size types like integer with custom counter
SELECT topn(topn_add_agg(int_column::text, 2), 3) FROM numbers WHERE int_column < 3;
 topn  
-------
 (2,6)
 (0,2)
(2 rows)

SELECT topn(topn_add_agg(int_column::text, 2), 4) FROM numbers WHERE int_column < 4;
 topn  
-------
 (2,6)
 (3,4)
(2 rows)

SELECT topn(topn_add_agg(int_column::text, 2), 4) FROM numbers WHERE int_column < 5;
 topn  
-------
 (2,6)
 (3,4)
(2 rows)

SELECT topn(topn_add_agg(int_column::text, 2), 4) FROM numbers WHERE int_column < 6;
 topn  
-------
 (5,7)
 (2,6)
(2 rows)

SELECT topn(topn_add_agg(int_column::text, 2), 4) FROM numbers;
 topn  
-------
 (5,7)
 (2,6)
(2 rows)

--check aggregates for variable length types like text
SELECT topn(topn_add_agg(text_column), 0) FROM strings WHERE text_column < '0';
 topn 
------
(0 rows)

SELECT topn(topn_add_agg(text_column), 1) FROM strings WHERE text_column < '1';
 topn  
-------
 (0,2)
(1 row)

SELECT topn(topn_add_agg(text_column), 2) FROM strings WHERE text_column < '2';
 topn  
-------
 (0,2)
 (1,1)
(2 rows)

SELECT topn(topn_add_agg(text_column), 3) FROM strings WHERE text_column < '3';
   topn    
-----------
 (2,15000)
 (0,2)
 (1,1)
(3 rows)

SELECT topn(topn_add_agg(text_column), 4) FROM strings WHERE text_column < '4';
   topn    
-----------
 (2,15000)
 (3,20)
 (0,2)
 (1,1)
(4 rows)

SELECT topn(topn_add_agg(text_column), 4) FROM strings WHERE text_column < '5';
   topn    
-----------
 (2,15000)
 (3,20)
 (4,6)
 (0,2)
(4 rows)

SELECT topn(topn_add_agg(text_column), 4) FROM strings WHERE text_column < '6';
   topn    
-----------
 (5,70000)
 (2,15000)
 (3,20)
 (4,6)
(4 rows)

SELECT topn(topn_add_agg(text_column), 4) FROM strings;
   topn    
-----------
 (5,70000)
 (2,15000)
 (3,20)
 (4,6)
(4 rows)

--check aggregates for cidr type
SELECT topn(topn_add_agg(cidr_column::text), 4) FROM cidr_table;
          topn          
------------------------
 (192.168.5.0/24,70000)
 (192.168.0.0/24,15000)
 (192.168.0.0/25,20)
 (192.168.1.0/24,6)
(4 rows)

SELECT topn(topn_add_agg(cidr_column::text, 2), 4) FROM cidr_table;
          topn          
------------------------
 (192.168.5.0/24,70000)
 (192.168.0.0/24,15000)
(2 rows)

--check aggregates for inet type
SELECT topn(topn_add_agg(inet_column::text), 4) FROM inet_table;
          topn           
-------------------------
 (10.1.2.3/32,70000)
 (192.168.2.1/24,15000)
 (192.168.100.128/23,20)
 (192.168.2.1/32,6)
(4 rows)

SELECT topn(topn_add_agg(inet_column::text, 2), 4) FROM inet_table;
          topn          
------------------------
 (10.1.2.3/32,70000)
 (192.168.2.1/24,15000)
(2 rows)

